scipy/sparse/tests/test_arithmetic1d.py

```diff
@@ -0,0 +1,337 @@
+"""Test of 1D arithmetic operations"""
+
+import pytest
+
+import numpy as np
+
+from scipy.sparse import coo_array, csr_array
+from scipy.sparse._sputils import isscalarlike
+
+
+spcreators = [coo_array, csr_array]
+math_dtypes = [np.int64, np.float64, np.complex128]
+
+
+def toarray(a):
+    if isinstance(a, np.ndarray) or isscalarlike(a):
+        return a
+    return a.toarray()
+
+@pytest.fixture
+def dat1d():
+    return np.array([3, 0, 1, 0], 'd')
+
+
+@pytest.fixture
+def datsp_math_dtypes(dat1d):
+    dat_dtypes = {dtype: dat1d.astype(dtype) for dtype in math_dtypes}
+    return {
+        sp: [(dtype, dat, sp(dat)) for dtype, dat in dat_dtypes.items()]
+        for sp in spcreators
+    }
+
+
+@pytest.mark.parametrize("spcreator", spcreators)
+class TestArithmetic1D:
+    def test_empty_arithmetic(self, spcreator):
+        shape = (5,)
+        for mytype in [
+            np.dtype('int32'),
+            np.dtype('float32'),
+            np.dtype('float64'),
+            np.dtype('complex64'),
+            np.dtype('complex128'),
+        ]:
+            a = spcreator(shape, dtype=mytype)
+            b = a + a
+            c = 2 * a
+            assert isinstance(a @ a.tocsr(), np.ndarray)
+            assert isinstance(a @ a.tocoo(), np.ndarray)
+            for m in [a, b, c]:
+                assert m @ m == a.toarray() @ a.toarray()
+                assert np.array_equal(m.dtype, mytype)
```

Comment:
style nit but here and elsewhere why not use `assert_equal` / `assert_allclose`?

Reply:
Somehow I got the message that the preference style choice with `pytest` is to use `assert` and then an expression. The "old-style" node api encouraged testing functions like `assert_equal`. With `pytest` stays closer to builtin python commands. It's also nice that `array_equal` is in the main `numpy` namespace so no additional imports are needed. But if SciPy has a style for testing that prefers the numpy testing functions, I can change it easily. Let me know.